homeassistant/components/fritz/sensor.py

```diff
@@ -0,0 +1,124 @@
+"""AVM FRITZ!Box binary sensors."""
+from __future__ import annotations
+
+import datetime
+import logging
+
+from fritzconnection.core.exceptions import FritzConnectionException
+
+from homeassistant.components.binary_sensor import BinarySensorEntity
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import DEVICE_CLASS_TIMESTAMP
+from homeassistant.core import HomeAssistant
+from homeassistant.util.dt import utcnow
+
+from .common import FritzBoxBaseEntity, FritzBoxTools
+from .const import DOMAIN, SENSOR_EXTERNAL_IP, SENSOR_UPTIME
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities
+) -> None:
+    """Set up entry."""
+    _LOGGER.debug("Setting up FRITZ!Box sensors")
+    fritzbox_tools = hass.data[DOMAIN][entry.entry_id]
+
+    if "WANIPConn1" in fritzbox_tools.connection.services:
+        # Only routers are supported at the moment
+
+        for sensor_type in [SENSOR_EXTERNAL_IP, SENSOR_UPTIME]:
+            async_add_entities(
+                [FritzBoxSensor(fritzbox_tools, entry.title, sensor_type)], True
+            )
+
+
+class FritzBoxSensor(FritzBoxBaseEntity, BinarySensorEntity):
+    """Define FRITZ!Box connectivity class."""
+
+    def __init__(
+        self, fritzbox_tools: FritzBoxTools, device_friendlyname: str, sensor_type: str
+    ) -> None:
+        """Init FRITZ!Box connectivity class."""
+        self._sensor_type = sensor_type
+        self._unique_id = (
+            f"{fritzbox_tools.unique_id}-{self._sensor_type.replace(' ', '_').lower()}"
+        )
+        self._name = f"{device_friendlyname} {self._sensor_type}"
+        self._is_available = True
+        self._last_value: str | None = None
+        self._state: str | None = None
+        super().__init__(fritzbox_tools, device_friendlyname)
+
+    @property
+    def name(self):
+        """Return name."""
+        return self._name
+
+    @property
+    def device_class(self) -> str | None:
+        """Return device class."""
+        if self._sensor_type == SENSOR_UPTIME:
+            return DEVICE_CLASS_TIMESTAMP
+        return None
+
+    @property
+    def icon(self):
+        """Return icon."""
+        if self._sensor_type == SENSOR_EXTERNAL_IP:
+            return "mdi:earth"
```

Comment:
Instead of including the logic inline make a `dict` that will provide the `icon`, `device_class` and function.

Reply:
Done